pyClient/zeth/joinsplit.py

```diff
@@ -601,46 +513,55 @@ def _compute_proof_hashes(proof_json: GenericProof) -> Tuple[bytes, bytes]:
         if key != "inputs":
             proof_elements.extend(proof_json[key])
     return (
-        sha256(encode_message_to_bytes(proof_elements)).digest(),
-        sha256(encode_pub_input_to_hash(proof_json["inputs"])).digest())
+        encode_message_to_bytes(proof_elements),
+        encode_message_to_bytes(proof_json["inputs"]))
 
 
 def sign_mix_tx(
+        signing_keypair: SigningKeyPair,  # Ephemeral signing key, tied to proof
         sender_eph_pk: EncryptionPublicKey,  # Ephemeral key used for encryption
         ciphertexts: List[bytes],  # Encyrpted output notes
         proof_json: GenericProof,  # Proof for the mix transaction
-        signing_keypair: SigningKeyPair  # Ephemeral signing key, tied to proof
 ) -> int:
+    """
+    Generate a Schnorr signature on the hash of the ciphertexts, proofs and
+    primary inputs. This is used to solve transaction malleability.  We chose
+    to sign the hash and not the values themselves for modularity (to use the
+    same code regardless of whether GROTH16 or PGHR13 proof system is chosen),
+    and sign the hash of the ciphers and inputs for consistency.
+    """
     assert len(ciphertexts) == constants.JS_INPUTS
-    # Hash the pk_sender and cipher-texts
-    sender_eph_pk_bytes = encode_encryption_public_key(sender_eph_pk)
-    ciphers = sender_eph_pk_bytes + ciphertexts[0] + ciphertexts[1]
-    hash_ciphers = sha256(ciphers).digest()
 
-    # Hash the proof
-    proof_hash, pub_inputs_hash = _compute_proof_hashes(proof_json)
+    # The message to sign consists of (in order):
+    #   - senders public encryption key
+    #   - ciphertexts
+    #   - proof elements
+    #   - public input elements
+    h = sha256()
+    h.update(encode_encryption_public_key(sender_eph_pk))
+    for ciphertext in ciphertexts:
+        h.update(ciphertext)
 
-    # Compute the joinSplit signature
-    return _sign(signing_keypair, hash_ciphers, proof_hash, pub_inputs_hash)
+    proof_hash, pub_inputs_hash = _compute_proof_hashes(proof_json)
+    h.update(proof_hash)
+    h.update(pub_inputs_hash)
+    message_digest = h.digest()
+    return sign(signing_keypair.sk, message_digest)
 
 
-def _compute_h_sig(
-        random_seed: bytes,
+def compute_h_sig(
         nf0: bytes,
         nf1: bytes,
         sign_pk: SigningPublicKey) -> bytes:
     """
-    Compute h_sig = blake2s(randomSeed, nf0, nf1, joinSplitPubKey)
+    Compute h_sig = sha256(nf0 || nf1 || sign_pk)
     Flatten the verification key
     """
-    blake_hash = blake2s()
-    blake_hash.update(random_seed)
-    blake_hash.update(nf0)
-    blake_hash.update(nf1)
-    for group_el in sign_pk:
-        for field_el in group_el:
-            blake_hash.update(field_el.n.to_bytes(32, byteorder="little"))
-    return blake_hash.digest()
+    h = sha256()
+    h.update(nf0)
+    h.update(nf1)
+    h.update(encode_vk_to_bytes(sign_pk))
```

Comment:
we should decide whether we call the verification key $vk$ or $pk$, here both notations are written which is a bit confusing.

Reply:
Yes.  I'll switch to vk everywhere.  That's what the key structures use now.

Reply:
Chaged